gpicker.1
.TH GPICKER 1 "January 2, 2010"
.SH NAME
gpicker \- fast (file)name chooser
.SH SYNOPSIS
.B gpicker
.RI [ options ]
.RI directory-path
.br
.SH DESCRIPTION
This manual page documents briefly the
.B gpicker
program.
.PP
.B gpicker
is a gtk+ program that allows you to quickly and conveniently pick
file in a (possibly very large) project. You type significant letters
of file name (typically from the start of words) and gpicker provides
you with a list of files you most likely mean to pick. The program
filters and orders project's list of files in real-time as you
type. It prints selected item (or items in multi-select mode) on
standard output.
.PP
You can use also
.B gpicker
to filter & sort arbitrary list of names by piping that list of names
to it's stdin and passing '-' instead of path.
.PP
.B gpicker
has a concept of project type which allows it to skip files that are
ignored by project's version control system. Default project type
invokes find to gather list of files. You can change find command line
via
.B GPICKER_FIND
environment variable.
.PP
.B gpicker
is typically used from programmer's editors or IDEs. VIM and Emacs
integration is included in
.B gpicker
package. See gpicker.el and gpicker.vim correspondingly.
There's also NetBeans plugin at
.B http://github.com/avsej/gpicker-netbeans
and gedit plugin at
.B http://github.com/yltsrc/gedit-gpicker
.PP
.B gpicker
can be used from other programs as well.
.B http://github.com/alk/supermegadoc
uses gpicker to pick names from documentation system index.
.br
.SH OPTIONS
This program follows the usual GNU command line syntax. A summary of options
is included below.
.TP
.B \-h, \-\-help
Show help options and quit
.TP
.B \-\-version
show version
.TP
.B \-t, \-\-project\-type
respect ignored files for given kind of VCS (default, git, bzr, hg,
guess, script). mlocate project type is used for loading file list
from mlocate database.

Type
.B script
will use
.B .gpicker-script
script located in project dir to read all relevant filenames. Script
must separate entries with zero character. It will be run with project
directory as current directory and should normally output relative
path. Not however that gpicker doesn't care if items output from
script are files and do exist.

Type
.B guess
will try to guess project type based on presence of directories and
files (such as
.B .gpicker-script
or
.B .git
).
.TP
.B \-\-disable\-bzr
disable autodetection of Bazaar project type
.TP
.B \-\-disable\-hg
disable autodetection of Mercurial project type
.TP
.B \-n, \-\-name\-separator
Set separator of filenames from stdin (\\0 is default).
Interprets \\0 as zero byte, \\n as new line, \\r as carriage return, \\t
as tab character.
.TP
.B \-d, \-\-dir\-separator
Sets separator of directory names from stdin (/ is default).
Interprets \\0 as zero byte, \\n as new line, \\r as carriage return, \\t
as tab character.
.TP
.B \-\-eat\-prefix
eat this prefix from names (./ is default)
.TP
.B \-m, \-\-multiselect
enable multiselect
.TP
.B \-l, \-\-left\-align
left align gpicker entries (default is right-align)
.TP
.B \-p, \-\-print\-pattern
print pattern if nothing matched
.TP
.B \-S, \-\-dont\-sort
dont sort result list
.TP
.B \-i, \-\-init\-filter
initial filter value
.TP
.B \-\-load\-stdin\-too
read additional filenames from stdin
.TP
.B \-I, \-\-output\-index
print selection index instead of value (implies \-S)
.TP
.B \-\-display=DISPLAY
X display to use
.br
.SH LICENSE
The gpicker is distributed under the GPLv3 license. See COPYING file
that comes with it. On Debian systems,
the complete text of the GNU General Public License
can be found in /usr/share/common-licenses/GPL-3.
.br
.SH SEE ALSO
gpicker-simple(1)
.PP
The README file that comes with gpicker.
.SH AUTHOR
The gpicker was written by Aliaksey Kandratsenka
.B <alk@tut.by>
with useful input and patches from folks in Ruby department of Altoros
Systems, Inc.
.B http://www.altoros.com/ruby_on_rails.html
.br


NEWS
gpicker NEWS -- history of user-visible changes.
Sat, 02 Jan 2010 20:57:04 +0200

version 2.0
===========

* gpicker.el doesn't hide gpicker-errors.log from user and unlinks it after gpicker
  invocation

* gpicker became even more faster. As part of performance work gpicker
  has been changed to truncate result list to 1000 items. Any gpicker
  optimizations would be invisible without that, as gtk+ list model
  isn't too fast. After more than year of using this feature I've
  never seen that truncated end of list. So it works great in
  practice.

* gpicker now filters & sorts results in separate thread. It
  gracefully abort current filtration when pattern is changed before
  filtration completion. This made UI extremely responsive, especially
  on huge filesets. I've tested this on list of my entire filesystem
  with more than 1M of files.

* Sergey Avseyev contributed VIM integration

* Sergey Avseyev contributed project types support for mercurial and
  bazaar. New project type 'guess' automagically detects project
  type.

* gpicker now has (out-of-tree) integrations with netbeans (by Sergey
  Avseyev) and gedit (by Yura Tolstik). See
  http://github.com/avsej/gpicker-netbeans and
  http://github.com/yltsrc/gedit-gpicker respectively.

* gpicker can now read names from stdin which enables it's use for
  choosing not just files, but anything.
  http://github.com/alk/supermegadoc uses this feature to pick index
  items (functions, types, ...) from various documentation systems.

* gpicker now has command-line options that control directory
  separator and name separator, as well as few other options that
  control name list loading.

  Directory separator option along with ability to read arbitrary
  names from stdin abstracts gpicker's notion of directory path to
  more general notion of namespace. In particular supermegadoc package
  uses erlang module name as namespace when navigating erlang
  documentation.

  Current version supports only single-char separators, yet. And
  there's special support for specifying new-line, zero byte, etc.

* it is now possible to enter directory/namespace filtration pattern
  after base pattern. Use '\', which is a special 'backwards'
  directory separator for such cases. 'bar\foo' is now equivalent to
  'foo/bar'. There are no command-line option for changing that
  separator.

  General use case for that feature is that you often start filtering
  by base name and look at what you get. You use directory name
  filtration usually only after you look at base name filter
  results. This feature helps you to avoid extra key presses precisely
  for that case. Just type '\' and start typing pattern for directory.

* gpicker (and it's Emacs integration) now supports multi-select. It
  is now possible to open .c and .h with one gpicker invocation. Emacs
  integration will automatically use split frame in that case.

* it's now possible to specify initial pattern. Emacs integration now
  uses word around point as initial pattern.

* gpicker file-list loading has been moved into separate thread, which
  allows gpicker UI to be responsive during file-list loading. You can
  now exit gpicker at any time.

* gpicker now supports building with (now obsolete) gtk-osx
  2.14. http://gtk-osx.org doesn't ship that gtk bundle anymore. But
  (possibly with minor ./configure.ac tweaks) it should build with
  latest gtk versions on OSX.

* it is now possible to control 'find' command line that gpicker uses to
  load list of files via GPICKER_FIND environment variable.

* WARNING: gpicker.el is not automagically installed anymore. Be
  careful to manually copy gpicker.el into correct place when
  installing or upgrading.

* Emacs integration now includes names of non-file backed buffers to
  select list and allows switching to them.

* gpicker now includes gpicker-simple which is non-interactive version
  of gpicker. Most importantly, gpicker-simple doesn't have any build
  dependencies.

  In default './configure' build gpicker-simple is just a hardlink to
  gpicker. But see Makefile.simple if you want to build
  dependency-free gpicker-simple. I think it even should be possible
  to build gpicker-simple on windows, with few tweaks (like providing
  missing getopt(3) implementation).

* gpicker.el now has (largely broken) ido integration (that depends
  only on gpicker-simple).

* gpicker now has initial support for matching UTF-8
  names. It is case-sensitive outside of ASCII though.

* gpicker now has manpages


gpicker.el
;;; gpicker.el --- gpicker integration for Emacs

;; Copyright (C) 2008 Aliaksey Kandratsenka

;; Author: Aliaksey Kandratsenka <alk@tut.by>
;; Version: 1.0
;; Keywords: autocompletion

;; This program is free software: you can redistribute it and/or
;; modify it under the terms of the GNU General Public License as
;; published by the Free Software Foundation; either version 3 of the
;; License, or (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
;; General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see
;; `http://www.gnu.org/licenses/'.

;;; Commentary
;;
;; Arrange loading of this file. One way is to add
;;
;;   (load "/usr/local/share/gpicker/gpicker.el")
;;
;; to your .emacs
;;
;; Run `gpicker-visit-project' to specify project directory.
;;
;; Open files by running `gpicker-find-file' or
;; `gpicker-find-file-other-window' or
;; `gpicker-find-file-other-frame'.
;;
;; I prefer to globally bind them to Super-f, C-x 4 Super-f and
;; C-x 5 Super-f correspondingly. You can do same by adding the following
;; elisp snippet to your .emacs
;;
;;   (global-set-key [8388710] 'gpicker-find-file)
;;   (global-set-key [?\C-x ?4 8388710] 'gpicker-find-file-other-window)
;;   (global-set-key [?\C-x ?5 8388710] 'gpicker-find-file-other-frame)
;;
;; or
;;
;;   (global-set-key (kbd "s-f") 'gpicker-find-file)
;;   (global-set-key (kbd "C-x 4 s-f") 'gpicker-find-file-other-window)
;;   (global-set-key (kbd "C-x 5 s-f") 'gpicker-find-file-other-frame)
;;
;; On most keyboards Super is a key with flag between Ctrl and Alt. On
;; mac keyboards it's Command key.
;;
;; Also check out gpicker-imenu and gpicker-isearch. As well
;; https://github.com/alk/supermegadoc/blob/master/supermegadoc.el

;;; Code

(eval-when-compile
  (require 'cl))

(require 'imenu)
(require 'ffap)

(defvar *gpicker-path* "gpicker")
(defvar *gpicker-simple-path* nil)
(defvar *gpicker-extra-args* '("--disable-bzr" ;; bzr builtin file listing is just too slow
                               "--load-stdin-too"
                               "--multiselect"))
(defvar *gpicker-project-dir* nil)
(defvar *gpicker-project-type* "guess")
(defvar *gpicker-errors-log* (expand-file-name "~/gpicker-errors.log"))
(defvar *gpicker-buffers-list* (expand-file-name "~/gpicker-buffers-list"))

(defun gpicker-delete-file (path)
  (condition-case e
      (delete-file path)
    (error nil)))

(defun gpicker-get-simple-path ()
  (or *gpicker-simple-path*
      (concat *gpicker-path* "-simple")))

(defun gpicker-grab-stdout (&rest call-process-args)
  (with-temp-buffer
    (unwind-protect (let ((status (apply #'call-process
                                         (car call-process-args)
                                         (and (file-exists-p *gpicker-buffers-list*)
                                              *gpicker-buffers-list*)
                                         (list (current-buffer) *gpicker-errors-log*)
                                         nil
                                         (cdr call-process-args))))
                      (if (eql status 0)
                          (buffer-string)
                        (message "%s exited with status %d" (car call-process-args) status)
                        (save-excursion
                          (with-current-buffer "*Messages*"
                            (goto-char (point-max))
                            (insert-file-contents *gpicker-errors-log*)))
                        nil))
      (progn
        (gpicker-delete-file *gpicker-errors-log*)
        (gpicker-delete-file *gpicker-buffers-list*)))))

(defun gpicker-pick (dir)
  (unless *gpicker-project-dir*
    (error "visit gpicker project via 'gpicker-visit-project first!"))
  (let ((gpicker-args (append *gpicker-extra-args*
                              (and *gpicker-project-type* (list "-t" *gpicker-project-type*))
                              (list dir)))
        (at-point (ffap-string-at-point)))
    (when (and at-point
               (> (string-bytes at-point) 0))
      (setq gpicker-args (list* "--init-filter" at-point gpicker-args)))
    (with-temp-file *gpicker-buffers-list*
      (let ((standard-output (current-buffer)))
        (dolist (b (buffer-list))
          (let ((name (buffer-name b)))
            (unless (or (eq (current-buffer) b)
                        (string= (substring name 0 1) " ")
                        (buffer-file-name b))
              (princ (buffer-name b))
              (princ "\0"))))))
    (unwind-protect (let ((rv (apply #'gpicker-grab-stdout
                                     *gpicker-path*
                                     gpicker-args)))
                      (and rv
                           (split-string rv "\0" t)))
      (discard-input))))

(defun gpicker-set-project-type (type)
  "Sets type of current gpicker project"
  (interactive (list (completing-read "Choose gpicker project type: "
                                      '("guess" "script" "git" "hg"
                                        "bzr" "default" "mlocate")
                                      nil t nil nil "guess")))
  (setq *gpicker-project-type* type))

(defun gpicker-visit-project (dir)
  "Selects current project directory for gpicker"
  (interactive "DSelect directory:")
  (cd dir)
  (setq *gpicker-project-type* "guess")
  (setq *gpicker-project-dir* (expand-file-name dir)))

(defun gpicker-call-find-function (find-function original-file)
  (let ((file (expand-file-name original-file *gpicker-project-dir*)))
    (if (file-exists-p file)
        (let ((revert-without-query (list (regexp-quote (abbreviate-file-name file)))))
          (funcall find-function file))
    (let ((b (get-buffer original-file)))
      (and b (display-buffer b))))))

(defun gpicker-do-find (find-function)
  (let ((files (gpicker-pick *gpicker-project-dir*)))
    (if (and (eql find-function #'find-file)
             (eql (length files) 2))
        (progn
          (gpicker-call-find-function #'find-file (car files))
          (gpicker-call-find-function #'find-file-other-window (cadr files)))
      (dolist (file files)
        (gpicker-call-find-function find-function file)))))

(defun gpicker-find-file ()
  (interactive)
  (gpicker-do-find #'find-file))

(defun gpicker-find-file-other-window ()
  (interactive)
  (gpicker-do-find #'find-file-other-window))

(defun gpicker-find-file-other-frame ()
  (interactive)
  (gpicker-do-find #'find-file-other-frame))

;;; ido integration

(defvar *gpicker-want-ido* nil
  "If ido integration is needed")

(when *gpicker-want-ido*
  (defvar *gpicker-hook-ido* t
    "Use gpicker for filtering ido results")

  (defadvice ido-set-matches-1 (around gpicker-ido-set-matches-1 activate)
    "Choose between the regular ido-set-matches-1 and gpicker-my-ido-match"
    (if *gpicker-hook-ido*
        (setq ad-return-value (gpicker-my-ido-match ido-text (ad-get-arg 0)))
      ad-do-it))

  (defun gpicker-ido-toggle ()
    "Toggle gpicker-ido integration"
    (interactive)
    (setq *gpicker-hook-ido* (not *gpicker-hook-ido*)))

  (defun gpicker-my-ido-match (str items)
    (setq items (reverse items))
    (with-temp-file *gpicker-buffers-list*
      (let ((standard-output (current-buffer)))
        (dolist (item items)
          (princ item)
          (princ "\0"))))
    (let ((args (list "-n\\0" str)))
      (when (or ido-rotate (string= str ""))
        (push "-S" args)) ;; dont_sort
      (let ((out (apply #'gpicker-grab-stdout
                        (gpicker-get-simple-path)
                        args)))
        (split-string out "\0" t)))))

;;; end of ido integration

(defun gpicker-complete-list (list &optional init-filter)
  (with-temp-file *gpicker-buffers-list*
    (let ((standard-output (current-buffer)))
      (dolist (name list)
        (princ name)
        (princ "\0"))))
  (unwind-protect (let ((rv (gpicker-grab-stdout *gpicker-path* "-l" "-d" "\01"
                                                 "--init-filter" (or init-filter (ffap-string-at-point))
                                                 "-")))
                    (and rv
                         (split-string rv "\0" t)))
    (discard-input)))

(defun alk-obarray-to-list (obarray)
  (let ((rv))
    (mapatoms (lambda (sym)
                (setq rv (cons (symbol-name sym) rv)))
              obarray)
    rv))

(defun gpicker-goto-tag ()
  (interactive)
  (find-tag (car (gpicker-complete-list (alk-obarray-to-list (tags-completion-table))))))

(defun gpicker-flatten-imenu (imenu-ilist &optional prefix)
  (let* ((prefix (or prefix ""))
         (list-of-lists (mapcar (lambda (pair)
                                  (let* ((snd (cdr pair))
                                         (item (car pair))
                                         (name (concat prefix item)))
                                    (if (listp snd)
                                        ;; if pair's cdr is list then it's sublist
                                        (gpicker-flatten-imenu snd (concat name "/"))
                                      ;; else it's marker
                                      (list (cons name snd)))))
                                imenu-ilist)))
    (apply #'append list-of-lists)))

(defun gpicker-imenu (&optional init-filter)
  (interactive)
  (let ((result t))
    (while (eq result t)
      (let* ((raw-imenu-alist (imenu--make-index-alist))
             (imenu-alist (gpicker-flatten-imenu raw-imenu-alist))
             (selected (car (gpicker-complete-list (mapcar #'car imenu-alist) init-filter))))
        (setq result (assoc selected imenu-alist))
        (when (equal result imenu--rescan-item)
          (imenu--cleanup)
          (setq result t))))
    (when result
      (imenu result))))

(defun gpicker-isearch ()
  (interactive)
  (let ((old-current-buffer (current-buffer)))
    (with-temp-file *gpicker-buffers-list*
      (let ((standard-output (current-buffer)))
        (princ (with-current-buffer old-current-buffer
                 (buffer-substring 1 (buffer-end 1)))))))
  (let ((rv (gpicker-grab-stdout *gpicker-path* "-IlP" "-d" "\\n" "-n" "\\n" "-")))
    (and (> (length rv) 0)
         (let ((line-num (1+ (string-to-number rv 10))))
           (goto-char (point-min))
           (forward-line (1- line-num))))))


(provide 'gpicker)
